Use arc-swap for swapping the pointer + other little things

vorner opened the pull request:
I originally wanted to do only https://github.com/slog-rs/slog/issues/222, but when I had the code opened, I added few more changes (hopefully improvements). These are each in separate commit, so if something of that is deemed unwanted, I can just drop them from the branch.

I have one further question, about the `new_from_arc` method. It leaks details about the internal implementation, which seems a bit wrong. I guess the original intention was to be able to create several interchangeable drains ‒ but after adding the `Clone` trait, it can be done in a more convenient way. Would it be OK to drop this method? It already causes a breaking change due to the change to arc-swap, so dropping it now would prevent a future breaking change on eg. arc-swap upgrade.

vorner committed:
Use arc-swap as the means to exchange things

vorner committed:
Code cleanups

Mostly using `dyn` where appropriate, but also some other small things.

vorner committed:
Traits for the drain

* They are cloneable.
* The drain can implement Default (by wrapping Discard)

vorner committed:
Delegate the is_enabled method

While this one is only an optimisation, it still makes sense to provide
it ‒ the inner drain may say no and then the caller can save some
performance.

dpc commented:
LGTM, but if they are breaking changes, we should bump the major version (which is OK). And if we're going to do that, I'm fine with any other changes you recommend.

vorner committed:
Remove the new_from_arc method

It leaks internal details and is no longer useful (interchangeable
Drains can be created by cloning the original).

vorner committed:
Changelog + version bump to 3.0.0

vorner commented:
OK, I've included the changes in this branch (except marking it in changelog, I don't know if you'd like to bundle any other changes before releasing).

vorner commented:
Thanks for merging.

Are there plans for some more changes before the release?

dpc commented:
I have none. I guess you're in charge now to decide. :D

vorner commented:
I don't see any need for further breaking changes at this time either, so I guess this would be nice to get released 😇.

dpc commented:
@vorner Go for it.

vorner commented:
I don't seem to have the upload rights, so it seems the `cargo publish` is up to you ;-).

dpc commented:
@vorner I've sent you an invite 10 hours ago. :)

vorner commented:
Thanks, I've managed to find it this time :tada: (the UI hides them quite well, there's nothing like an email or notification bar anywhere).
